shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Scripts.hs

```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Cardano.Ledger.ShelleyMA.Scripts (Script) where
+
+import Cardano.Binary
+  ( Annotator,
+    FromCBOR (..),
+    ToCBOR (..),
+    encodeListLen,
+  )
+import qualified Cardano.Ledger.Core as Core
+import qualified Cardano.Ledger.Crypto as CryptoClass
+import Cardano.Ledger.Era (Era)
+import qualified Cardano.Ledger.Shelley as Shelley
+import Cardano.Ledger.ShelleyMA (MaryOrAllegra, ShelleyMAEra)
+import Cardano.Ledger.ShelleyMA.Timelocks
+  ( Timelock,
+    ValidityInterval,
+    hashTimelockScript,
+    validateTimelock,
+  )
+import Data.Coders (decodeRecordSum)
+import Data.Typeable (Typeable)
+import Data.Word (Word8)
+import GHC.Generics (Generic)
+import GHC.Records
+import NoThunks.Class
+import Shelley.Spec.Ledger.Scripts (MultiSig)
+import Shelley.Spec.Ledger.Tx
+  ( ValidateScript (..),
+    hashMultiSigScript,
+    validateNativeMultiSigScript,
+  )
+
+data Script era
+  = ScriptMSig (MultiSig era)
+  | ScriptTimelock (Timelock era)
+  deriving (Show, Typeable, Eq, Generic)
+
+instance Typeable era => NoThunks (Script era)
+
+instance Typeable era => ToCBOR (Script era) where
+  toCBOR = \case
+    ScriptMSig s -> encodeListLen 2 <> toCBOR (0 :: Word8) <> toCBOR s
+    ScriptTimelock s -> encodeListLen 2 <> toCBOR (1 :: Word8) <> toCBOR s
+
+instance Era era => FromCBOR (Annotator (Script era)) where
```

Comment:
Note that this serialisation is not correct. https://jira.iohk.io/browse/CAD-2145 covers the task of making the serialisation compatible, and @redxaxder has suggested a means by which this can be achieved.

Reply:
(But in theory, we can go to a testnet without this)